### PR TITLE
Add plan-all option

### PR DESCRIPTION
**Pull request by jocgir**
I submitted this one few weeks ago, but it was refused because plan-all will fail on a stack with dependencies that has never been deployed.

As suggested, I added a mechanism to give user an explicit message in the case where the plan fail because its dependencies were not satisfied.

Also, as specified in #194, I relied on `RunModules` instead of `RunModulesReverseOrder` for both plan-all and output-all.

### CODE REVIEW

File: configstack/stack.go [hunk 1]
```diff
@@ -25,6 +26,40 @@ func (stack *Stack) String() string {
...
+			if len(stack.Modules[n].Dependencies) > 0 && strings.Contains(output, ": Resource 'data.terraform_remote_state.") {
+				terragruntOptions.Logger.Printf(
+					"%v contains dependencies to %v and refers to remote state, "+
+						"you may have to apply your changes in the dependencies prior running terragrunt plan-all.\n",
+					stack.Modules[n].Path,
+					stack.Modules[n].Config.Dependencies.Paths,
+				)
+			}
+		}
+	}
```
Comment (brikis98) [inline configstack/stack.go]: This for-loop should be moved into its own method called something like `summarizePlanAllErrors`.

File: configstack/stack.go [hunk 2]
```diff
@@ -25,6 +26,40 @@ func (stack *Stack) String() string {
...
+	err := RunModules(stack.Modules)
+
+	// We inspect the error streams to give an explicit message if the plan failed because there were references to
+	// remote states. `terraform plan` will fail if it tries to access remote state from dependencies and the plan
+	// has never been applied on the dependency.
+	for n, o := range errorStreams {
+		output := o.String()
+		if strings.Contains(output, "Error running plan:") {
+			terragruntOptions.Logger.Println(output)
+			if len(stack.Modules[n].Dependencies) > 0 && strings.Contains(output, ": Resource 'data.terraform_remote_state.") {
```
Comment (brikis98) [inline configstack/stack.go]: It's easy to forget to include a module in the `Dependencies` list, so perhaps that check is superfluous, and searching for the `data.terraform_remote_state` error is enough.

File: configstack/stack.go [hunk 3]
```diff
@@ -25,6 +26,40 @@ func (stack *Stack) String() string {
...
+		module.TerragruntOptions.ErrWriter = &errorStreams[n]
+	}
+
+	// Start the plan on each module
+	err := RunModules(stack.Modules)
+
+	// We inspect the error streams to give an explicit message if the plan failed because there were references to
+	// remote states. `terraform plan` will fail if it tries to access remote state from dependencies and the plan
+	// has never been applied on the dependency.
+	for n, o := range errorStreams {
```
Comment (brikis98) [inline configstack/stack.go]: `i` is a more common name for an index variable. `errorStream` is probably an easier to understand name than `o`.

File: configstack/stack.go [hunk 4]
```diff
@@ -25,6 +26,40 @@ func (stack *Stack) String() string {
...
+		output := o.String()
+		if strings.Contains(output, "Error running plan:") {
+			terragruntOptions.Logger.Println(output)
```
Comment (brikis98) [inline configstack/stack.go]: Won't this print out the entire plan error message again?

File: configstack/stack.go (hunk 3 above)
Comment (jocgir) [inline configstack/stack.go]: Yes, your are right, I have tendency to use small variables and rename them later.
Fixed

File: configstack/stack.go (hunk 2 above)
Comment (jocgir) [inline configstack/stack.go]: OK

**Comment by jocgir** [comment]
The goal of plan-all is mainly to evaluate the impact of code modification on an already applied infrastructure. When we submit a PR on our infrastructure, our CI apply a plan-all to let us see the impact and we have to confirm if we wish to apply them.

It can still work on new infrastructure, but not if there are true dependencies.

File: configstack/stack.go (hunk 4 above)
Comment (jocgir) [inline configstack/stack.go]: No, the error message is not printed initially. Is is only printed there.

But some messages are

**Comment by jocgir** [comment]
Regarding #80, I agree with you. The xxx-all outputs are difficult to read mainly due to the multithreading. Capturing the output (stdout and stderr) for each and summarize at the end would give a more readable output.

Also, there is a problem with the exit code. We should recuperate exit code (err) for each module and summarize them with the output.

I especially have a problem with `terraform plan -detailed-exitcode`. It returns 2 if there are modification, but terragrunt exit with 1 because it is not able to parse the error summary.

`[terragrunt] 2017/04/26 19:07:07 Unable to determine underlying exit code, so Terragrunt will exit with error code 1`

Also, I don't know if it is a bug in terraform, but for some specific cases, when we use 

```
<= module.domain.data.aws_route53_zone.main
    caller_reference:          "<computed>"
    comment:                   "<computed>"
    name:                      "dev.xxx.com."
    private_zone:              "true"
    resource_record_set_count: "<computed>"
    tags.%:                    "<computed>"
    vpc_id:                    "<computed>"
    zone_id:                   "<computed>"

Plan: 0 to add, 0 to change, 0 to destroy.
[terragrunt] [/terragrunt/infra] 2017/04/26 19:44:10 Module /terragrunt/infra has finished with an error: exit status 2```

As you can see, we receive `Plan: 0 to add, 0 to change, 0 to destroy.`, but terraform still exit with status 2.

File: README.md [hunk 5]
```diff
@@ -743,13 +743,25 @@ cd root
...
+To see the currently applied outputs of all of the subfolders, you can use the `output-all` command:
 
 ```
 cd root
 terragrunt output-all
 ```
 
+Finally, if you make some changes to your project, you could evaluate the impact by using `plan-all` command:
+
+Note: It is important to realize that you could get errors running `plan-all` if you have dependencies between your projects.
```
Comment (brikis98) [inline README.md]: s/if you have dependencies between your projects/if you have dependencies between your projects and some of those dependencies haven't been applied yet/

File: README.md [hunk 6]
```diff
@@ -743,13 +743,25 @@ cd root
...
+
+_Ex: If you import a state file from one of your dependencies, `plan-all` will not generate or update that state file, resulting
+in possible errors in the dependant project._
```
Comment (brikis98) [inline README.md]: I'd rephrase the example to: "If module A depends on module B and module B hasn't been applied yet, then plan-all will show the plan for B, but exit with an error when trying to show the plan for A"

**Comment by brikis98** [comment]
> I especially have a problem with terraform plan -detailed-exitcode. It returns 2 if there are modifications, but terragrunt exit with 1 because it is not able to parse the error summary.

Please file a bug for that. AFAIK, that used to work, so we may have a regression that broke that functionality.

> As you can see, we receive Plan: 0 to add, 0 to change, 0 to destroy., but terraform still exit with status 2.

Never seen that before. Is there a simple case to repro that with? Does it only happen with Terragrunt or does the same problem happen when using Terraform directly?

**Comment by jocgir** [comment]
The fact that exit code 2 is given when it should not is a terraform problem. But terragrunt still has problem reporting it and convert the exit code 2 (changes) to exit code 1 because of this:
```go
type MultiError struct {
	Errors []error
}

func (err MultiError) Error() string {
	errorStrings := []string{}
	for _, err := range err.Errors {
		errorStrings = append(errorStrings, err.Error())
	}
	return fmt.Sprintf("Encountered the following errors:\n%s", strings.Join(errorStrings, "\n"))
}
```

The extra formatting make the error difficult to analyse at the end of the program.

File: README.md (hunk 6 above)
Comment (jocgir) [inline README.md]: Done

File: README.md (hunk 5 above)
Comment (jocgir) [inline README.md]: Done

**Comment by brikis98** [comment]
> The fact that exit code 2 is given when it should not is a terraform problem. But terragrunt still has problem reporting it and convert the exit code 2 (changes) to exit code 1 because of this:

Ah, I see. Worth filing a separate bug to track it. Not clear what exit code to use with multiple `plan` commands though, but better logging would definitely be a good idea.

**Comment by brikis98** [comment]
OK, LGTM. Merging. Thank you!

**Comment by brikis98** [comment]
https://github.com/gruntwork-io/terragrunt/releases/tag/v0.12.13
